### PR TITLE
Add testing regarding SparseAdam state_dicts

### LINKED ISSUE #116507: Add testing regarding SparseAdam state_dicts
### 🚀 The feature, motivation and pitch

SparseAdam, unlike the other optimizers, does not test any state_dict functionality. All other optims call _test_state_dict, but due to SparseAdam's only handling sparse gradients, it does not easily fall into the architecture that already exists.

We should consider giving it its own closure that would update parameters like in `rosenbrock` and creating special case tests for it.

### Alternatives

Leave SparseAdam state_dict handling untested

### Additional context

_No response_

cc @vincentqb @jbschlosser @albanD @crcrpar @mruberry @ZainRizvi

### PR DESCRIPTION
**Pull request by jovianjaison**
Summary:
- Updated SparseAdam to run test_state_dict_deterministic unit test.
- Made gradients sparse while keeping weights dense in the above test.

Test Plan:
- Ran test_optim.py locally.

Fixes #116507

### CODE REVIEW

**Review by janeyx99 — APPROVED** [review]

File: test/test_optim.py [hunk 1]
```diff
@@ -1327,6 +1327,11 @@ def fwd_bwd(optim, w, b, i):
             optim.zero_grad()
             loss = (w.mv(i) + b).pow(2).sum()
             loss.backward()
+            if optim.__class__.__name__ == "SparseAdam":
```
[the remark this answers is not in the document]
Comment (janeyx99) [reply test/test_optim.py:1330]: Please modify this before merging. Thanks for the fix!
